src/hotspot/share/utilities/globalDefinitions_gcc.hpp

```diff
@@ -139,10 +139,21 @@ inline int g_isfinite(jdouble f)                 { return isfinite(f); }
 #endif // _LP64
 
 // gcc warns about applying offsetof() to non-POD object or calculating
-// offset directly when base address is NULL. Use 16 to get around the
-// warning. The -Wno-invalid-offsetof option could be used to suppress
-// this warning, but we instead just avoid the use of offsetof().
-#define offset_of(klass,field) (size_t)((intx)&(((klass*)16)->field) - 16)
+// offset directly when base address is NULL. The -Wno-invalid-offsetof
+// option could be used to suppress this warning, but we instead just
+// avoid the use of offsetof().
+//
+// FIXME: This macro is complex and rather arcane. Perhaps we should
+// use offsetof() instead, with the invalid-offsetof warning
+// temporarily disabled.
+#define offset_of(klass,field)                          \
+[]() {                                                  \
+  char space[sizeof (klass)] ATTRIBUTE_ALIGNED(16);     \
```

Comment:
Is 16 over-aligned on 32bit platforms? Too bad https://git.openjdk.org/jdk/pull/11446 isn't approved yet, else you could use alignas(std::max_align_t). (I wouldn't object if you were forward-looking here, though can't speak for anyone else.)

Reply:
I'll leave it at 16. This is a bug fix, and I don't want it to have a dependency on [JDK-8297912](https://bugs.openjdk.org/browse/JDK-8297912), a P4 Enhancement.

Reply:
OK.